appserver/web/war-util/src/test/java/org/glassfish/web/loader/WebappClassLoaderTest.java

```diff
@@ -0,0 +1,178 @@
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+package org.glassfish.web.loader;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+
+import org.apache.naming.resources.FileDirContext;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class WebappClassLoaderTest {
+
+    private static CyclicBarrier lock;
+    private static ExecutorService executor;
+    private static File junitJarFile;
+
+    @BeforeClass
+    public static void setup() throws URISyntaxException {
+        // Run 3 methods at the same time, and make the pool large enough to increase
+        // the chance of a race condition
+        lock = new CyclicBarrier(3);
+        executor = Executors.newFixedThreadPool(60);
+
+        // Fetch any JAR to use for classloading
+        junitJarFile = new File(Test.class.getProtectionDomain().getCodeSource().getLocation().toURI());
+    }
+
+    @AfterClass
+    public static void shutdown() {
+        if (executor != null) {
+            executor.shutdownNow();
+        }
+    }
+
+    @Test
+    public void check_findResourceInternalFromJars_thread_safety() throws Exception {
+        final ClassLoader classLoader = this.getClass().getClassLoader();
+        final WebappClassLoader webappClassLoader = new WebappClassLoader(classLoader, null);
+        webappClassLoader.start();
+        webappClassLoader.setResources(new FileDirContext());
+
+        CompletableFuture<Exception> result = new CompletableFuture<>();
+
+        // Create the tasks, and have them each run at the same time
+        // using the cyclic barrier
+        Runnable lookupTask = waitAndDo(lock, result, () -> lookup(classLoader, webappClassLoader));
+        Runnable addTask = waitAndDo(lock, result, () -> add(classLoader, webappClassLoader));
+        Runnable closeTask = waitAndDo(lock, result, () -> webappClassLoader.closeJARs(true));
+
+        try {
+            // Run the methods at the same time
+            for (int i = 0; i < 100; i++) {
+                executor.execute(addTask);
+                executor.execute(lookupTask);
+                executor.execute(closeTask);
+            }
+            // Check to see if any completed exceptionally
+            Exception ex = result.get(200, TimeUnit.MILLISECONDS);
+            if (ex != null) {
+                throw ex;
+            }
+        } catch (TimeoutException ex) {
+            // Success!
+        } finally {
+            webappClassLoader.close();
+        }
+    }
```

Comment:
I would add this, but it depends on the purpose of the test - if deadlock was acceptable result, then this would only break the test:
```
        List<Runnable> notExecutedYet = executor.shutdownNow();
        assertTrue("some jobs never executed: " + notExecutedYet.size(), notExecutedYet.isEmpty());
```
It will fail if the executor pool is so small that it can execute some of tasks and some must wait in a queue. If tasks are not ending, the pool will not be capable to run more ...

So now we have 60 threads in pool for 300 tasks (3*100). Easy.